File: src/main/java/com/github/kpavlov/jreactive8583/netty/pipeline/ReconnectOnCloseListener.java
package com.github.kpavlov.jreactive8583.netty.pipeline;

import com.github.kpavlov.jreactive8583.client.Iso8583Client;
import io.netty.channel.Channel;
import io.netty.channel.ChannelFuture;
import io.netty.channel.ChannelFutureListener;
import org.slf4j.Logger;

import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;

import static org.slf4j.LoggerFactory.getLogger;

public class ReconnectOnCloseListener implements ChannelFutureListener {

    private final Logger logger = getLogger(ReconnectOnCloseListener.class);

    private final Iso8583Client client;
    private final int reconnectInterval;
    private final AtomicBoolean disconnectRequested = new AtomicBoolean(false);
    private final ScheduledExecutorService executorService;

    public ReconnectOnCloseListener(Iso8583Client client, int reconnectInterval, ScheduledExecutorService executorService) {
        this.client = client;
        this.reconnectInterval = reconnectInterval;
        this.executorService = executorService;
    }

    public void requestReconnect() {
        disconnectRequested.set(false);
    }

    public void requestDisconnect() {
        disconnectRequested.set(true);
    }

    @Override
    public void operationComplete(ChannelFuture future) {
        final Channel channel = future.channel();
        logger.debug("Client connection was closed to {}", channel.remoteAddress());
        channel.disconnect();
        scheduleReconnect();
    }

    public void scheduleReconnect() {
        if (!disconnectRequested.get()) {
            logger.trace("Failed to connect. Will try again in {} millis", reconnectInterval);
            executorService.schedule(
                    client::connectAsync,
                    reconnectInterval, TimeUnit.MILLISECONDS);
        }
    }
}

File: src/main/kotlin/com/github/kpavlov/jreactive8583/netty/pipeline/ReconnectOnCloseListener.kt
package com.github.kpavlov.jreactive8583.netty.pipeline

import com.github.kpavlov.jreactive8583.client.Iso8583Client
import io.netty.channel.ChannelFuture
import io.netty.channel.ChannelFutureListener
import org.slf4j.LoggerFactory
import java.util.concurrent.Callable
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean

public open class ReconnectOnCloseListener(
    private val client: Iso8583Client<*>,
    private val reconnectInterval: Int,
    private val executorService: ScheduledExecutorService,
) : ChannelFutureListener {
    private val logger = LoggerFactory.getLogger(ReconnectOnCloseListener::class.java)
    private val disconnectRequested = AtomicBoolean(false)

    public fun requestReconnect() {
        disconnectRequested.set(false)
    }

    public fun requestDisconnect() {
        disconnectRequested.set(true)
    }

    public override fun operationComplete(future: ChannelFuture) {
        val channel = future.channel()
        logger.debug("Client connection was closed to {}", channel.remoteAddress())
        channel.disconnect()
        scheduleReconnect()
    }

    public fun scheduleReconnect() {
        if (!disconnectRequested.get()) {
            logger.trace("Failed to connect. Will try again in {} millis", reconnectInterval)
            executorService.schedule(
                Callable { client.connectAsync() },
                reconnectInterval.toLong(),
                TimeUnit.MILLISECONDS,
            )
        }
    }
}
